src/passes/transform/cse.cpp

```diff
@@ -0,0 +1,174 @@
+#include "coreir/passes/transform/cse.h"
+
+using namespace std;
+using namespace CoreIR;
+
+// Desired CoreIR iterators/functional methods
+// breadth-first tree iterator on wireables
+// depth-first tree iterator on wireables
+// leaf-iterator on wireables (Only single bits)
+// breadth-first iteration on types
+// depth-first iterator on types
+// leaf iterator on types 
+// 'All' operator with a predicate
+// 'Any' operator with a predicate
+// 'Map' operator
+// 'Filter' operator (both a modify and a return new)
+
+
+//Some possible (untested) implementations of some of the functions
+template <typename Cont, typename Pred>
+Cont filterOver(const Cont &cont, Pred predicate) {
```

Comment:
I don't have access to the header file, so I can't tell but are these supposed to be externally available free functions? Or meant to be just for internal use to this CSE pass? If they're external then the should probably be in some more generic file (ala `util.hpp` or `functional.hpp`);  otherwise they should be in anonymous namespace:

```c++
namespace {

template<...>
Cont filterOver(...) {
  ...
}

...

}  // namespace
```